### PR TITLE
proposal for shared services

### PR DESCRIPTION
**Pull request by philbrookes**

### CODE REVIEW

File: apbs/use-shared-services-across-namespaces.md
Comment (maleck13) [inline apbs/use-shared-services-across-namespaces.md]: might want to add a little more to the definition (the service need to support being shared) For instance at the moment I don't think there is anyway to share the sync service

File: apbs/use-shared-services-across-namespaces.md
Comment (maleck13) [inline apbs/use-shared-services-across-namespaces.md]: A service that can be used across multiple namespaces

File: apbs/use-shared-services-across-namespaces.md
Comment (maleck13) [inline apbs/use-shared-services-across-namespaces.md]: Change Keycloak to Authentication Service

File: apbs/use-shared-services-across-namespaces.md
Comment (maleck13) [inline apbs/use-shared-services-across-namespaces.md]: Should we specify that these are specifically provision parameters and maybe given an example of what they look like

File: apbs/use-shared-services-across-namespaces.md
Comment (maleck13) [inline apbs/use-shared-services-across-namespaces.md]: > and create a configuration as required.

And run through the rest of the APB as though the instance had finished provisioning the actual running instance

**Comment by maleck13** [comment]
@philbrookes I think this is a good start, however I think we should also cover how a binding will work for a shared service. For instance, if you had a shared keycloak and then performed a binding against that serviceinstance that is pointing to a service in a different namespace. I don't think there will be too many difficulties but I think we should cover it in the proposal

**Review by wei-lee — COMMENTED** [review]
@philbrookes @maleck13 I have a few questions. If this is enabled, is it possible to have this situation:

* In namespace A, fh-sync service is bound to KC in namepsace B
* In namespace A, metrics service (or maybe another service) is bound to KC in namespace C

So we end up with 2 services are bound to completely different KC instances.

First of all, should we allow this to happen?

Secondly, if this is the case, how should we deal with it in the auth SDK? At the moment, the auth SDK only works across services that are protected by the same KC that the user is authenticated with.

**Comment by maleck13** [comment]
> In namespace A, metrics service (or maybe another service) is bound to KC in namespace C

@wei-lee  I don't think this will happen.
When provisioning the second KC it there will be a new serivceInstance created in namespace A that refers to the KC in namespace B. When asking for a binding within a namespace, it uses the service instance in that namespace as a reference. So  it will be a binding to the keycloak that the serviceintance is associated with. In this case the one in namespace B.

@philbrookes  I think we have to consider what may happen though if a developer deletes the serviceinstance in Namespace A, the deprovision playbook needs to be intelligent enough to only remove the realm and secrets / configmaps and not the the actual running KC instance

File: apbs/use-shared-services-across-namespaces.md
Comment (philbrookes) [reply apbs/use-shared-services-across-namespaces.md]: I've attached an image below this paragraph showing what they might look like, is that what you meant?

**Comment by wei-lee** [comment]
@maleck13 ah, right. Gotcha. There will still be a KC instance in the namespace A which should be used for binding by other services. But this instance maybe a pointer to a instance in another namespace. Is it right to say?

**Comment by maleck13** [comment]
@wei-lee there will be a serviceinstance in namespace A yes that serviceinstance is "backed" by the KC in  namespace B. Any bindings that happen in namespace A will be directed to KC in namespace B

**Comment by maleck13** [comment]
Also needs to cover in detail how a deprovision on a shared service should behave. (spoke with Phil about this on a call). Proposal should cover how each action that can be performed against a regular instance will now work against a shared service

File: apbs/use-shared-services-across-namespaces.md
Comment (david-martin) [inline apbs/use-shared-services-across-namespaces.md]: Are we limiting users to services already deployed to a namespace? What about services running outside of kubernetes?

File: apbs/use-shared-services-across-namespaces.md
Comment (david-martin) [inline apbs/use-shared-services-across-namespaces.md]: suggesting "adoption of Mobile Services" instead of "adoption of the Mobile Platform"

File: apbs/use-shared-services-across-namespaces.md
Comment (david-martin) [reply apbs/use-shared-services-across-namespaces.md]: Image gives a good example :+1:

File: apbs/use-shared-services-across-namespaces.md
Comment (david-martin) [inline apbs/use-shared-services-across-namespaces.md]: Should/If

File: apbs/use-shared-services-across-namespaces.md
Comment (david-martin) [inline apbs/use-shared-services-across-namespaces.md]: Is there upstream work/proposals/roadmap items that would address this?
Good to call this out, but it's not specific to the shared services scenario. Its a potential difficulty with any APB.

**Review by david-martin — CHANGES_REQUESTED** [review]
@philbrookes few comments. Nothing major

File: apbs/use-shared-services-across-namespaces.md
Comment (maleck13) [inline apbs/use-shared-services-across-namespaces.md]: Do we want to add deprovision and bind goals here?

File: apbs/use-shared-services-across-namespaces.md
Comment (maleck13) [inline apbs/use-shared-services-across-namespaces.md]: I think we need to make this a little bit clearer, I know what you mean but could be hard to parse.
So essentially if the resources created during the provision are present in the namespace then it will deprovision them otherwise it will only remove service specific items (such as the clients etc) and not the actual running pods

**Review by maleck13 — CHANGES_REQUESTED** [review]
2 small comments but looking good now

**Review by maleck13 — APPROVED** [review]

**Comment by maleck13** [comment]
@david-martin you ok to give it another look over and approve if you are happy

**Review by david-martin — APPROVED** [review]

**Comment by david-martin** [comment]
I'm happy with the changes after feedback :+1:  Merging

**Pull request merged**
